[week7] 인가(Authorization) 구현하기

etff opened the pull request:
## 🎯 요구사항
### 인가 구현하기

 * [x]  1. 비밀번호 암호화
 * [x]  2. 인가 기능을 위해 Spring Security에 적용 

---
### 기타 요구사항
* [x]  테스트 커버리지 100%를 달성해야 합니다.
* [x]  모든 API 테스트를 통과해야 합니다.
* [x]  모든 E2E 테스트를 통과해야 합니다.

etff committed:
add spring security

Spring Security는 Spring 기반의 애플리케이션의 보안(인증과 권한, 인가 등)을 담당하는 스프링 하위 프레임워크입니다.
Spring Security는 '인증'과 '권한'에 대한 부분을 Filter 흐름에 따라 처리하고 있습니다.

see also
https://spring.io/projects/spring-security

etff committed:
feature: add security config

add custom user authentication

etff committed:
feature: add authentication filter

etff committed:
feature: add preAuthorized check

etff committed:
test: delete product without access token

etff committed:
refactor: add password encoder

비밀번호 암호화를 위해 BCryptPasswordEncoder를 도입합니다.Spring Security에서 단방향 암호화를 위한 기능을 제공합니다.
BcryptPasswordEncoder는 BCrypt 해시 함수를 사용해 비밀번호를 암호화를 진행합니다.

etff committed:
refactor: add user encode password match when update, register

etff committed:
fix: update fail when not found user

etff committed:
fix: url authentication permit

etff committed:
fix: user update fail response

etff commented:
테스트를 보면 업데이트시 존재하지 않는 자원에 403이 되기를 기대하고 있는데
NotFound와 Forbidden의 어떤 차이로 구분하면 좋을지 잘 모르겠습니다. 어떤 경우에 Forbidden을 쓰면 좋을까요?

그리고 존재하지 않는 사용자 삭제시 응답이 Not Found로 유지하면되는지 궁금합니다.

hannut91 commented:
자신의 정보만 수정할 수 있도록 변경되면서, `Not Found` 에러가 나기 전에 이미 `Forbidden`에러가 발생해서 그런 것 같습니다.
사실 사용자를 찾을 수 없다면, 내 자신이 없는 경운데 이런 일은 발생하지 않죠 ㅎㅎ

만약에 관리자 권한이 생기고, 관리자는 다른 사람의 정보도 삭제할 수 있다면 그 때는 `Not Found`에러가 발생할 수 있겠죠

etff committed:
doc: add javadoc

etff committed:
refactor: add security character filter

add character utf-8 filter

etff committed:
feature: add no args constructor

etff committed:
test: check user id match

etff committed:
feature: check user id match

etff committed:
refactor: add User constructor

etff committed:
feature: add user id match when update user

etff committed:
test: add update with authentication

etff committed:
feature: add access denied exception

etff committed:
fix: change change user to forbidden response

etff committed:
test: add uncheck method

etff committed:
refactor: remove lombok generated code

etff committed:
refactor: remove unused exception response

etff committed:
test: add user authentication creation

etff committed:
refactor: move id check before find user

etff committed:
fix: change auth token with not existed id

etff committed:
test: user authentication credential return null

hannut91 commented on app/src/main/java/com/codesoom/assignment/application/UserService.java:
```diff
@@ -16,26 +19,39 @@
-    public User updateUser(Long id, UserModificationData modificationData) {
-        User user = findUser(id);
+    public User updateUser(Authentication authentication, Long id, UserModificationData modificationData) {
+        Long userDetailId = (Long) authentication.getPrincipal();
```
인증을 처리하는 책임을 Controller에게 갖게 하고, 여기서는 수정하는데 필요한 로직만 하는 것은 어떨까요?

사용자의 정보를 수정하는 기능과 `Authentication`과 결합이 되어서 `Authentication`이 변경되면, 여기도 변경이 될 수 있을 것 같아서 결합도가 증가한 것 같습니다

etff replied:
좋은 방법인 것같습니다. 

추가로 궁금한점은 
해결해야하는 요구사항은 id 비교이기때문에, 차라리 인증객체를 보내는 것이 아닌 id만 보내는 건 어떨까요?
트레이너님의 생각도 궁금합니다.

hannut91 replied:
말씀해주신대로 `id`만 보내는게 좋을 것 같습니다!

hannut91 commented on app/src/test/java/com/codesoom/assignment/controllers/UserControllerTest.java:
```diff
@@ -112,8 +142,39 @@ void updateUserWithValidAttributes() throws Exception {
+    }
 
-        verify(userService).updateUser(eq(1L), any(UserModificationData.class));
+    @Test
+    void updateUserWithAccessDeniedToken() throws Exception {
+        mockMvc.perform(
+                patch("/users/999")
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .content("{\"name\":\"TEST\",\"password\":\"test\"}")
+                        .header("Authorization", "Bearer " + VALID_TOKEN)
+        )
+                .andExpect(status().isForbidden());
```
사용자가 자신의 정보가 아닌 다른 사람의 정보를 업데이트 하려고할 때를 테스트해주신 것 같아요.
토큰이 접근 권한이 없는 토큰이라는 것이 잘 보이지가 않네요. `VALID_TOKEN`은 `userId`가 1인 사용자의 토큰이라는 것을 기억해야만 이 테스트 코드를 이해할 수 있을 것 같아요. 더 좋은 방법이 있을까요? 음 저는

```java
mockMvc.perform(patch("/users/" + userId + 10))
```

이렇게해서 아이디가 다르다는 것을 표현해야 하나? 고민이 되었어요. 맘에들지는 않네요

etff replied:
지금보니 URL에 있는 id 만 갖고 코드를 이해하기 힘드네요. 좀더 수정을 해보겠습니다.

etff committed:
refactor: change url path name

테스트시 업데이트 대상자와 인증토큰 대상자가 다름을 표시하기위해
변수명과 Path를 변경합니다.

etff committed:
test: move wrong parameter test when user update

테스트 순서를 바꿨습니다.
인증 객체 id와 path variable id 테스트를 먼저 진행하여 id가 9999로 고정되었기때문에
wrong parameter 테스트를 with others 상태에서 진행하는 것이 됩니다.

테스트 목적은 parameter 테스트이기 때문에 path와 인증객체가 일치하는 상태에서
paramter만 비교하는게 맞다고 생각합니다.

디버깅을 해보면 업데이트시 정상인 경우 1번, path variable이 다른 경우가 9999로 두번 들어옵니다.

hannut91 commented on app/src/test/java/com/codesoom/assignment/controllers/UserControllerTest.java:
```diff
@@ -169,10 +172,10 @@ void updateUserWithInvalidToken() throws Exception {
     void updateUserWithAccessDeniedToken() throws Exception {
         mockMvc.perform(
-                patch("/users/999")
+                patch("/users/2")
```
`patch("/users/" + USER2)`
이렇게 하고 싶으셨던 거죠? ㅎㅎ 좋네요

etff replied:
`patch("/users/" + USER2)`

좀 더 다른 사용자임을 나타내게 되었네요 👍

hannut91 reviewed (commented):
테스트 순서가 올바르지 않았군요 ㅎㅎ 좋네요

말씀하신대로 더 이해하기 쉽게 순서를 변경하는게 좋겠군요

etff commented on app/src/main/java/com/codesoom/assignment/config/SecurityConfig.java:
```diff
@@ -0,0 +1,58 @@
+        Filter authenticationFilter = new JwtAuthenticationFilter(
+                authenticationManager(), authenticationService);
+        Filter authenticationErrorFilter = new AuthenticationErrorFilter();
+        CharacterFilter characterFilter = new CharacterFilter();
+
+        http
+                .addFilterBefore(characterFilter.getFilter(), CsrfFilter.class)
+                .csrf()
```
uft-8 인코딩 설정을 위해 필터를 잡아주었습니다. 구글링을 해보니
SpringSecurity 사용시. CsrfFilter 앞에 CharacterEncodingFilter를 놓아야 한다는 내용이 있어 
csrf 앞에 두었습니다.

see also
http://millky.com/@origoni/post/1057

etff commented on app/src/main/java/com/codesoom/assignment/config/SecurityConfig.java:
```diff
@@ -0,0 +1,58 @@
+                        new HttpStatusEntryPoint(HttpStatus.UNAUTHORIZED));
+    }
+
+    @Override
+    public void configure(WebSecurity web) throws Exception {
+        web.ignoring()
+                .antMatchers("/h2-console/**/**");
+    }
```
h2-console에 보면 이미지 같은 정보가 나오지않고 있어
정적자원 설정으로 WebSecurity를 적용하였습니다.

좀 더 구글링을 해보고, 리뷰 내용을 보니 HttpSecurity에서 FrameHeader Option을 설정하는게 더 맞는거 같네요

etff commented on app/src/main/java/com/codesoom/assignment/application/UserService.java:
```diff
@@ -16,26 +19,39 @@
+        }
+        final User user = findUser(id);
+
+        final String encodedPassword = passwordEncoder.encode(modificationData.getPassword());
+        final User source = User.of(modificationData.getName(), encodedPassword);
```
- PasswordEncoder가 User에 들어가면 
도메인 영역이 외부에 의존하게 되기때문에 따로 User에 changePassword같은 메서드를 생성하지 않았습니다.

hannut91 replied:
고민이 되는 부분이군요. 
이전에는 `User`에게 비밀번호를 변경해 라는 `changePassword`라는 메시지를 전달하면, 알아서 변경이 되었는데, 지금은 비밀번호를 어떻게 인코딩하고 사용자가 비밀번호를 어떻게 저장하는지 `UserService`에서 다 알고 있는 것 같네요.
설계는 트레이드 오프라는데 이런걸 보고 하는 말인지 고민이 됩니다

etff commented on app/src/main/java/com/codesoom/assignment/application/UserService.java:
```diff
@@ -16,26 +19,39 @@
+        Long userDetailId = (Long) authentication.getPrincipal();
+        if (!id.equals(userDetailId)) {
+            throw new AccessDeniedException(userDetailId);
+        }
+        final User user = findUser(id);
+
+        final String encodedPassword = passwordEncoder.encode(modificationData.getPassword());
+        final User source = User.of(modificationData.getName(), encodedPassword);
 
-        User source = mapper.map(modificationData, User.class);
         user.changeWith(source);
```
Object Mapper가 간단한 객체 변환이면 유용한데,  암호화같은 사용자 요구사항이 들어가기 시작하면
쓰기 힘들어지는 것 같습니다.
`changePassword`를 만드는 방법도 있지만, 이미 쓰고있는 `changeWith`를 사용하고 싶었습니다. 
`changePassword`를 만들면 `changeWith`는 이름 변경하는 역할밖에 하지 못하기 때문입니다.

etff commented on app/src/main/java/com/codesoom/assignment/application/AuthenticationService.java:
```diff
@@ -5,30 +5,54 @@
-        if (!user.authenticate(password)) {
+        if (!passwordEncoder.matches(password, user.getPassword())) {
             throw new LoginFailException(email);
         }
```
Id 비교처럼 User도메인에 암호화된 비밀번호만 전달해서 원래 암호비교를 하고싶었는데
passwordEncoder가 랜덤으로 암호화해서 그렇게하지못했습니다. 이게 최선이라고 생각하는데 아쉽습니다.

etff replied:
userdetails 같은 정보가 들어오면 거기에 credentail 정보를 사용해서 user 도메인에 
비밀번호를 확인하라 메시지를 줄 수 있겠네요

etff committed:
refactor: add authentication access filter

기존에 사용자 권한체크를 service layer에서 진행하였습니다.
사용자 권한 체크를 `SecurityConfig` 이 책임을 갖게하고,
User Controller 와 User Service 는 요청을 받아 계정을 처리하는 역할만 갖게하였습니다.

etff committed:
refactor: change name guard to authentication guard

인증 처리를 하는 Guard 클래스 의미를 명확하게 변경하였습니다.

etff committed:
refactor: URL에 사용자 이름을 넣어 어떤 사용자가 등장하는지 명확하게 변경

`/users/1` 으로 표현하던 URL에서 `/user/ + user_id` 로 표현하여
여러 사용자가 등장했을때 magic number로 헷갈리는 것을 방지하였습니다.

etff committed:
refactor: enable header frame option

기존에 h2 database console이 spring security 설정으로 인해 사용이 제한되었습니다.
h2-console의 frame option을 사용하여, X-Frame-Options deny 옵션을 사용할 수 있게 변경하였습니다.

see also
> https://developer.mozilla.org/ko/docs/Web/HTTP/Headers/X-Frame-Options

etff committed:
refactor: add app config

application의 bean 등록을 위한 설정 클래스를 생성.

etff committed:
refactor: add cors filter

기존의 cross origin 의 책임을 controller이 담당하고 있었습니다.
보안의 책임을 SecurityConfig 가 같고있어 filter에서 cors를 설정하도록 변경하였습니다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/application/AuthenticationGuard.java:
```diff
@@ -0,0 +1,16 @@
+import org.springframework.security.core.Authentication;
+import org.springframework.stereotype.Component;
+
+@Component
+public class AuthenticationGuard {
+
+    public boolean checkIdMatch(Authentication authentication, Long id) {
+        if (authentication == null) {
+            return false;
+        }
+        Long userDetailId = (Long) authentication.getPrincipal();
+        return id.equals(userDetailId);
+    }
+}
```
`AuthenticationGuard `라는 객체가 `checkIdMatch`라는 것을 할 수 있는데, 이 기능을 하는데 필요한 데이터를 객체가 가지고 있지 않고 모두 인자로 받고 있습니다. 그러다보니 사실 이 메서드를 static으로 선언해도 아무런 문제가 없어보이네요

이 클래스는 util 클래스로 보는게 좋을까요?

etff replied:
util 클래스는 아니라고 생각해요. 

Util성  클래스를 만들때는 어느 객체나 서비스구간에서 혹시 책임을 다하지 않는가 고민해 볼 필요가 있는것 같네요.
static이 나쁘다는 이유는 아니고요. 인증하는 과정에서 협력하고 있는 객체이기 때문에 
굳이 util성 클래스로 만들 필요는 없다고 생각합니다.

hannut91 replied:
협력하고 있긴 한데, 이 객체가 가지고 있는 데이터가 하나도 없어서, util 클래스처럼 보인다고 생각했습니다 ㅎㅎ

etff committed:
refactor: add MyPasswordEncoder

UserService에서 사용자의 정보를 등록, 갱신하는 과정에서
비밀번호 암호화 작업을 하는 책임을 지고 있습니다.

비밀번호를 암호화를 담당하는 객체와 협력을 하도록 변경했습니다.
너무 많은 정보가 드러나지 않도록 인터페이스를 이용하였습니다.

[@johngrib 님이 소개한 인터페이스를 이용한 객체간 협력](https://github.com/CodeSoom/spring-week6-assignment-1/pull/14#issuecomment-790609410)

etff commented:
지난주에 배운 인터페이스를 응용해서 UserService에서 암호화 객체와 협력하도록 변경했습니다.

hannut91 commented:
좋습니다! 저는 패스워드가 패스워드 데이터 자체도 가지고 있고, encoder도 가지고 있어서 스스로 encode할 수 있도록 하고 싶었는데 쉽지 않네요.

```java
public class Password {
    private String password;
    private PasswordEncoder encoder;
    
    public Password(PasswordSupplier supplier, PasswordEncoder encoder) {
        this(supplier.getPassword(), encoder);
    }

    public Password(String password, PasswordEncoder encoder) {
        this.password = password;
        this.encoder = encoder;
    }

    public String getPassword() {
        return this.password;
    }

    public void encode() {
        this.password = encoder.encode(this.password);
    }

    public Boolean isEqual(String password) {
        return this.encoder.matches(password, this.password);
    }
}
```

etff commented:
> 좋습니다! 저는 패스워드가 패스워드 데이터 자체도 가지고 있고, encoder도 가지고 있어서 스스로 encode할 수 있도록 하고 싶었는데 쉽지 않네요.
> 
> ```java
> public class Password {
>     private String password;
>     private PasswordEncoder encoder;
>     
>     public Password(PasswordSupplier supplier, PasswordEncoder encoder) {
>         this(supplier.getPassword(), encoder);
>     }
> 
>     public Password(String password, PasswordEncoder encoder) {
>         this.password = password;
>         this.encoder = encoder;
>     }
> 
>     public String getPassword() {
>         return this.password;
>     }
> 
>     public void encode() {
>         this.password = encoder.encode(this.password);
>     }
> 
>     public Boolean isEqual(String password) {
>         return this.encoder.matches(password, this.password);
>     }
> }
> ```

비밀번호만 바꿔주는 도메인 서비스 기능이라면 이런 방식으로 구현하는 것도 좋을 것 같습니다. 
구현시에는 PasswordEncoder를 bean으로 등록했기때문에 의존성 주입에서 힘들 것 같네요

hannut91 commented:
책임을 어떤 객체에게 할당할지 고민하신 부분이 인상깊습니다 ㅎㅎ 덕분에 많이 배웠습니다 감사합니다 ㅎㅎㅎ
이번주 고생많으셨습니다!

etff commented:
7주차 회고입니다.
https://itcoin.tistory.com/564

트레이너님과 코드리뷰를 통해 프로젝트 구조에 대해 고민을 하는 좋은 시간이었습니다.
